main.py

```diff
@@ -0,0 +1,109 @@
+from typedb.client import *
+
+def highestVaccines():
+
+    with session.transaction(TransactionType.READ) as read_transaction:
+
+        answer_iterator = read_transaction.query().match_group_aggregate('match $p isa person, has certificate_no $no, has country $c; get $p, $no, $c; group $c; count;')
+
+        stats = []
+
+        for answer in answer_iterator:
+            stats.append([answer._numeric._int_value, answer._owner._value])
+
+        stats.sort(reverse = True)
+
+        print("\nTop 3 countries with most Vaccinations:")
+        for i in range(3):
+            print(stats[i][1], ": ", stats[i][0])
+
+def vaccineByAge(lower_limit, upper_limit):
+
+    with session.transaction(TransactionType.READ) as read_transaction:
+
+        answer_iterator = read_transaction.query().match_group_aggregate(f'match $p isa person, has age >= {lower_limit}, has age <= {upper_limit}; $m isa manufacturer, has vaccine_name $v; $vcc (person: $p, manufacturer: $m) isa vaccinator; get $p, $v; group $v; count;')
+
+        print(f"\nVaccines given to age group {lower_limit} - {upper_limit} years:")
+
+        for answer in answer_iterator:
+            print(answer._owner._value, ": ", answer._numeric._int_value)
+
+def vaccineByCountry(country):
+
+    with session.transaction(TransactionType.READ) as read_transaction:
+
+        country_iterator = read_transaction.query().match('match $p isa person, has country $c; get $c;')
+
+        countries = []
+        for country_pres in country_iterator:
+            name = country_pres.get("c")
+            countries.append(name.get_value())
+
+        if country not in countries:
+            print("No entries for the given country!")
+            return
+
+        answer_iterator = read_transaction.query().match_group_aggregate(f'match $p isa person, has country = "{country}"; $m isa manufacturer, has vaccine_name $v; $vcc (person: $p, manufacturer: $m) isa vaccinator; get $p, $v; group $v; count;')
+
+        print(f"\nVaccines provided in {country}:")
+        for answer in answer_iterator:
+            print(answer._owner._value, ": ", answer._numeric._int_value)
+
+def avgAgeByCountry():
+
+    with session.transaction(TransactionType.READ) as read_transaction:

```

Comment:
Good improvement - however these functions are now all dependent on having background state in the form of being called from within the scope of a `with ... as session`. All of these functions would be much more usable if you either
- passed in the session as a parameter explicitly
- passed in the transaction as a parameter explicitly, and created it within the `while` loop

or something similar to one of those options.

Reply:
If the functions took the session as an argument, then they wouldn't rely on the existence of `with ... as session`. e.g.
```py
def avgAgeByCountry(session):
```

This would make the functions much more self-contained - which is generally a good idea.

Reply:
I removed that "with" statement for the session opening, and just assigned it directly.
Do I still need to pass the session in the functions?

Reply:
I would still recommend passing the session into the functions - as a general rule, it's good if any state that the function has to interact with is passed in as a parameter - this way anyone reading the code can understand how that function works without having to look elsewhere in the code to see the state that the function relies on.